### DISCUSSION TITLE
Is there any way to extract a NeogmaInstance from a NeogmaModel?

### DISCUSSION THREAD

**Discussion by Ansis100** [themetalfleece/neogma#96] [Q&A]
I have a function which returns a Neogma model with a neogma instance dependency injection, for example:
```ts
function User(neogma: Neogma): NeogmaModel<UserProperties, UserRelatedNodes, UserMethods>;
```

Is there any way to extract a NeogmaInstance from this? I attempted to do it like this:
```ts
export type ModelToInstance<
  FunctionT extends (neogma: Neogma) => ModelT,
  ModelT extends NeogmaModel<
    PropertiesT,
    RelatedNodesT
  > = ReturnType<FunctionT>,
  PropertiesT extends Neo4jSupportedProperties = Record<string, any>,
  RelatedNodesT extends object = object,
> = NeogmaInstance<PropertiesT, RelatedNodesT>;

type UserInstance = ModelToInstance<typeof User
```

Which doesn't throw errors but also doesn't fully work. For some reason the `ReturnType<FunctionT>` part completely ignores the `NeogmaModel` return type and creates some weird generic type that doesn't even look like a `NeogmaInstance` and doesn't have any property definitions.

**Comment by Ansis100** [comment]
I just noticed that it probably doesn't detect the property definitions because the generic parameter `PropertiesT` is optional with an `object` initializer, so it defaults to `any` for everything. I still have no idea how to make it detect the properties from the type parameters of the passed model type.

**Comment by Ansis100** [comment]
In the meantime I figured it was just easier to export the instance types from the model definitions.

**Reply by marisb11** [reply]
```ts
export type ExtractedInstanceType<AnyModel > = AnyModel extends (
  neogma: Neogma,
) => NeogmaModel<infer Properties, infer RelNodes, infer Methods, any> // any as Statics placeholder 
  ? NeogmaInstance<Properties, RelNodes, Methods>
  : never;
```

**Reply by themetalfleece** [reply]
Sorry @Ansis100, it seems like I totally missed this discussion.
That's what I thought, the Instance type is independent from the neogma dependency, right?
Are you still having an issue or have you figured this out? :)
